dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/PreCheckDataRelationsValidationHook.java

```diff
@@ -146,177 +166,354 @@ public void validateEvent( ValidationErrorReporter reporter, Event event )
+                allCOsExist = false;
             }
         }
-        else if ( ENROLLMENT.equals( trackerType ) )
+        return allCOsExist;
+    }
+
+    private boolean hasNoAttributeCategoryOptionsSet( Event event )
+    {
+        return StringUtils.isBlank( event.getAttributeCategoryOptions() );
+    }
+
+    private Set<String> parseCategoryOptions( Event event )
+    {
+
+        String cos = StringUtils.strip( event.getAttributeCategoryOptions() );
+        if ( StringUtils.isBlank( cos ) )
         {
-            if ( uid.isPresent() && !ValidationUtils.enrollmentExist( ctx, uid.get() ) )
-            {
-                reporter.addError( relationship, E4012, trackerType.getName(), uid.get() );
-            }
+            return Collections.emptySet();
+        }
+
+        return TextUtils
+            .splitToArray( cos, TextUtils.SEMICOLON );
+    }
+
+    /**
+     * Validates that the event program has the default category combo if no AOC
+     * or COs are given or if a default AOC is given.
+     *
+     * @param reporter validation error reporter
+     * @param event event to validate
+     * @param program event program from the preheat
+     * @return return true if event program is default with valid aoc and co
+     *         combinations
+     */
+    private boolean validateDefaultProgramCategoryCombo( ValidationErrorReporter reporter, Event event,
+        Program program )
+    {
+        if ( hasNoAttributeOptionComboSet( event ) &&
+            hasNoAttributeCategoryOptionsSet( event ) &&
+            !program.getCategoryCombo().isDefault() )
+        {
+            reporter.addError( event, TrackerErrorCode.E1055 );
+            return false;
         }
-        else if ( EVENT.equals( trackerType ) )
+        CategoryOptionCombo aoc = reporter.getValidationContext().getBundle().getPreheat()
+            .getCategoryOptionCombo( event.getAttributeOptionCombo() );
+        if ( hasAttributeOptionComboSet( event ) &&
+            aoc != null && aoc.getCategoryCombo().isDefault() &&
+            !program.getCategoryCombo().isDefault() )
         {
-            if ( uid.isPresent() && !ValidationUtils.eventExist( ctx, uid.get() ) )
-            {
-                reporter.addError( relationship, E4012, trackerType.getName(), uid.get() );
-            }
+            reporter.addError( event, TrackerErrorCode.E1055 );
+            return false;
         }
+
+        return true;
     }
 
-    // TODO: This method needs some love and care, the logic here is very hard
-    // to read.
-    protected void validateEventCategoryCombo( ValidationErrorReporter reporter,
-        Event event, Program program )
+    private boolean hasAttributeOptionComboSet( Event event )
     {
-        TrackerImportValidationContext context = reporter.getValidationContext();
-        TrackerPreheat preheat = reporter.getValidationContext().getBundle().getPreheat();
+        return !hasNoAttributeOptionComboSet( event );
```

Comment:
I find it more readable. Most of the checks use `hasNoAttributeOptionComboSet`. That's why I created it using `hasNo..` instead of `has`. So we don't have to do `!has...` all the time. Reading a "positive" check is easier, than negating in my head. There is a check that uses its negation `hasAttributeOptionComboSet` which to me is easier to read than `!hasNoAttributeOptionComboSet()`(that makes my head spin 😅).